File: src/main/scala/challenge/c2021/c2021_10/c2021_10_08.scala
package challenge.c2021.c2021_10

import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

/** [[https://leetcode.com/problems/implement-trie-prefix-tree/]] */
class c2021_10_08 extends AnyWordSpec with Matchers {

  class Trie() {
    private var wordEnd = false
    private val data = Array.fill[Option[Trie]](26)(None)

    def insert(word: String): Unit =
      word.map(_ - 'a').foldLeft(this) { (curr, i) =>
        if (curr.data(i).isEmpty)
          curr.data(i) = Some(new Trie())
        curr.data(i).get
      }.wordEnd = true

    def search(word: String): Boolean =
      find(word).exists(_.wordEnd)

    def startsWith(prefix: String): Boolean =
      find(prefix).isDefined

    private def find(prefix: String): Option[Trie] =
      prefix
        .map(_ - 'a')
        .foldLeft[Option[Trie]](Some(this)) { (curr, i) =>
          curr.flatMap(_.data(i))
        }
  }

  /**
   * Your Trie object will be instantiated and called as such:
   * var obj = new Trie()
   * obj.insert(word)
   * var param_2 = obj.search(word)
   * var param_3 = obj.startsWith(prefix)
   */

  "Example 1" in {
    val trie = new Trie()
    trie.insert("apple")
    trie.search("apple") shouldBe true
    trie.search("app") shouldBe false
    trie.startsWith("app") shouldBe true
    trie.insert("app")
    trie.search("app") shouldBe true
  }

  private def intToWord2000(i: Int): String = {
    val sb = new StringBuilder(2000)
    val str = Array.fill[Char](5)('a')
    i.toString.map(_ - '0' + 'a').map(_.toChar).zipWithIndex.foreach { case (c, i) => str(i) = c }
    for (_ <- 0 to 400) sb.appendAll(str)
    sb.toString()
  }

  "performance" should {
    "insert 300x2000 strings" in {
      val trie = new Trie()
      (1 to 300)
        .map(intToWord2000)
        .foreach(trie.insert)
    }

    // 3 sec 798 ms; 3.6 GiB
    "insert 10kx2k search 10kx2k startsWith 10kx2k" in {
      val trie = new Trie()
      val words = (1 to 10000).map(intToWord2000)
      words.foreach(trie.insert)
      words.map(trie.search).foreach(_ shouldBe true)
      words.map(trie.startsWith).foreach(_ shouldBe true)
    }
  }
}

File: rust/2021-10/src/c2021_10_08.rs
#![allow(dead_code)]
/// 208. Implement Trie (Prefix Tree)
/// =================================
///
/// A [trie] (pronounced as "try") or __prefix tree__ is a tree data structure
/// used to efficiently store and retrieve keys in a dataset of strings.
/// There are various applications of this data structure, such as autocomplete and spellchecker.
///
/// Implement the Trie class:
///
/// - `Trie()`                            Initializes the trie object.
/// - `void insert(String word)`          Inserts the string `word` into the trie.
/// - `boolean search(String word)`       Returns `true` if the string `word` is in the trie
///                                       (i.e., was inserted before), and `false` otherwise.
/// - `boolean startsWith(String prefix)` Returns `true` if there is a previously inserted string `word`
///                                       that has the prefix `prefix`, and `false` otherwise.
///
/// [trie]: https://en.wikipedia.org/wiki/Trie
///
/// __Constraints:__
///
/// - `1 <= word.length, prefix.length <= 2000`
/// - `word` and `prefix` consist only of lowercase English letters.
/// - At most `30_000` calls __in total__ will be made to `insert`, `search`, and `startsWith`.
///
/// https://leetcode.com/problems/implement-trie-prefix-tree/
#[derive(Debug, Default)]
struct Trie {
    word_end: bool,
    data: [Option<Box<Self>>; 26],
}
impl Trie {
    fn new() -> Self {
        Self::default()
    }

    fn insert(&mut self, word: String) {
        let mut curr = self;
        for i in word.bytes().map(|b| (b - b'a') as usize) {
            if curr.data[i].is_none() {
                curr.data[i] = Some(Box::new(Self::new()));
            }
            curr = curr.data[i].as_mut().unwrap();
        }
        curr.word_end = true;
    }

    fn search(&self, word: String) -> bool {
        self.find(word).map(|t| t.word_end).unwrap_or(false)
    }

    fn starts_with(&self, prefix: String) -> bool {
        self.find(prefix).is_some()
    }

    fn find(&self, prefix: String) -> Option<&Self> {
        let mut curr = self;
        for i in prefix.bytes().map(|b| (b - b'a') as usize) {
            if curr.data[i].is_none() {
                return None;
            } else {
                curr = curr.data[i].as_ref().unwrap();
            }
        }
        Some(curr)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn example1() {
        let mut t = Trie::new();
        t.insert("apple".to_string());
        assert!(t.search("apple".to_string()));
        assert!(!t.search("app".to_string()));
        assert!(t.starts_with("app".to_string()));
        t.insert("app".to_string());
        assert!(t.search("app".to_string()));
    }

    fn i32_to_word(i: i32) -> String {
        String::from_utf8(
            format!("{:0<5}", i)
                .into_bytes()
                .into_iter()
                .map(|b| b - b'0' + b'a')
                .collect(),
        )
        .unwrap()
    }

    mod performance {
        use super::*;

        //#[ignore]
        #[test]
        fn insert_00300x2000() {
            let mut t = Trie::new();
            for i in 1..=300 {
                let w = i32_to_word(i).repeat(400);
                t.insert(w);
            }
        }

        /// my Trie: 11 sec 408 ms; 3.7 GiB
        // #[ignore]
        #[test]
        fn insert_10kx2k_search_10kx2k_starts_with_10kx2k() {
            let mut t = Trie::new();
            let mut words = vec![];
            for i in 1..=10000 {
                words.push(i32_to_word(i).repeat(400));
            }

            for w in words.clone() {
                t.insert(w);
            }
            for w in words.clone() {
                assert!(t.search(w));
            }
            for w in words.clone() {
                assert!(t.starts_with(w));
            }
        }
    }

    // mod produce_test_cases {
    //     use super::*;
    //
    //     const LEN: i32 = 300;
    //
    //     #[test]
    //     fn produce_commands() {
    //         let mut commands = vec!["Trie"];
    //         for _ in 1..=LEN {
    //             commands.push("insert");
    //         }
    //         println!("{:?}", commands);
    //     }
    //
    //     #[test]
    //     fn produce_params() {
    //         let mut words = vec![vec![]];
    //         for i in 1..=LEN {
    //             let w = i32_to_word(i).repeat(400);
    //             words.push(vec![w]);
    //             //t.insert(w);
    //         }
    //         println!("{:?}", words);
    //     }
    // }
}
